src/eval/reduction.rs

```diff
@@ -0,0 +1,1828 @@
+
+                        if rest.is_nil() || !end.is_nil() {
+                            Control::Error(expr, env)
+                        } else {
+                            Control::Return(
+                                arg1,
+                                env,
+                                cont_witness.intern_named_cont(
+                                    ContName::NewerCont,
+                                    store,
+                                    Continuation::Unop {
+                                        operator: Op1::U64,
+                                        continuation: cont,
+                                    },
+                                ),
+                            )
+                        }
+                    } else if head == c.comm.ptr() {
+                        let (arg1, end) = car_cdr_named!(ConsName::ExprCdr, &rest)?;
+
+                        if rest.is_nil() || !end.is_nil() {
+                            Control::Error(expr, env)
+                        } else {
+                            Control::Return(
+                                arg1,
+                                env,
+                                cont_witness.intern_named_cont(
+                                    ContName::NewerCont,
+                                    store,
+                                    Continuation::Unop {
+                                        operator: Op1::Comm,
+                                        continuation: cont,
+                                    },
+                                ),
+                            )
+                        }
+                    } else if head == c.char.ptr() {
+                        let (arg1, end) = car_cdr_named!(ConsName::ExprCdr, &rest)?;
+
+                        if rest.is_nil() || !end.is_nil() {
+                            Control::Error(expr, env)
+                        } else {
+                            Control::Return(
+                                arg1,
+                                env,
+                                cont_witness.intern_named_cont(
+                                    ContName::NewerCont,
+                                    store,
+                                    Continuation::Unop {
+                                        operator: Op1::Char,
+                                        continuation: cont,
+                                    },
+                                ),
+                            )
+                        }
+                    } else if head == c.eval.ptr() {
+                        if rest.is_nil() {
+                            return Ok((Control::Error(expr, env), None));
+                        }
+                        let (arg1, more) = car_cdr_named!(ConsName::ExprCdr, &rest)?;
+
+                        if rest.is_nil() {
+                            Control::Error(expr, env)
```

Comment:
Apparently `car_cdr_named!` can modify `rest`, likely walking over a `cons` pointer.
So the first check for `nil` assures that `eval` can't have 0 arguments. Then, after consuming two pointers, we want the (modified) `rest` to point to `nil` because `eval` can't have more than 2 arguments.

Reply:
Indeed, the macro hides the receiver of the `.car_cdr_named()` call, which is the mutable (and mutated) pointer in this case.

Reply:
Just a quick note: I am going to completely ignore all discussion about the existing code which just happens to appear in the diff of this PR. Otherwise it will be impossible to ever complete the current work. I mention this only to set expectations and explain why I am not engaging with these discussions here/now.